File: src/alpha_sortedness.c
#include <stdio.h>
#include <stdlib.h>

int* genAlphaSortedArray(float alpha, int n){
    int i;
    int* array = calloc(n, sizeof(int));
    for(i = 0; i < n; i++){
        array[i] = i;
    }
    int num_replace = (int)((1 - alpha) * n);
    for(i = 0; i < num_replace; i++){
        int index = (int)(rand() % n);
        if(array[index] == n + 1){
            i--;
            continue;
        }
        array[index] = n + 1;
    }
    return array;
}

int* getRandomSubArrayLocal(int* array, int gamma, int n){
    int index = (int)(rand() % (n - gamma));
    return &array[index];
}

float calcAlphaSortedness(int* sub_array, int gamma, int n){
    (void)n;
    int i;
    int num_unsorted = 0;
    for(i = 0; i < gamma - 1; i++){
        if(sub_array[i] >= sub_array[i + 1]){
            num_unsorted++;
        }
    }
    return (1.0 * (gamma - num_unsorted)) / gamma;
}

int main(int argc, char* argv[]){
    time_t t;
    int i, rep;
    srand((unsigned int) time(&t));

    if(argc != 7){
        printf("Usage is %s array_size alpha beta gamma array_alpha num_repetitions\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    int n = atoi(argv[1]);
    float alpha = atof(argv[2]);
    int beta = atoi(argv[3]);
    int gamma = atoi(argv[4]);
    float alpha_prime = atof(argv[5]);
    int num_repetitions = atoi(argv[6]);

    int* array = genAlphaSortedArray(alpha_prime, n);
    int successes = 0;
    for(rep = 0; rep < num_repetitions; rep++){
        float total_alpha = 0;
        for(i = 0; i < beta; i++){
            int* sub_array = getRandomSubArrayLocal(array, gamma, n);
            total_alpha += calcAlphaSortedness(sub_array, gamma, n);
        }
        float approx_alpha = total_alpha / beta;
        if(approx_alpha >= alpha){
            successes += 1;
        }
    }
    
    printf("Succeeded %d out of %d times (%4f)\n", successes, num_repetitions, (1.0 * successes) / num_repetitions);

    free(array);
    return 0;
}

File: src/alpha_sortedness.py
import random
import sys

array = []

def genAlphaSortedArray(alpha, n):
    global array
    array = range(0, n)
    num_replace = int((1 - alpha) * n)
    for i in xrange(0, num_replace):
        index = random.randint(0, n-1)
        if index == 0:
            array[index] = array[index + 1] + 1
        else:
            array[index] = array[index - 1] - 1

def getRandomSubArray(gamma):
    global array
    index = random.randint(0, len(array) - gamma - 1)
    return array[index:index + gamma]

def calcAlphaSortedness(sub_array):
    num_unsorted = 0
    for i in xrange(0, len(sub_array) - 1):
        if sub_array[i] > sub_array[i + 1]:
            num_unsorted += 1
    return float(len(sub_array) - num_unsorted) / len(sub_array)

def main():
    random.seed()

    if(len(sys.argv) != 5):
        print "Usage is %s array_size alpha beta gamma" % sys.argv[0]
        return
    n = int(sys.argv[1])
    alpha = float(sys.argv[2])
    beta = int(sys.argv[3])
    gamma = int(sys.argv[4])

    genAlphaSortedArray(alpha, n)

    total_alpha_sortedness = 0
    for i in xrange(0, beta):
        total_alpha_sortedness += calcAlphaSortedness(getRandomSubArray(gamma))
        print total_alpha_sortedness
    approximate_alpha_sortedness = float(total_alpha_sortedness) / beta
    if approximate_alpha_sortedness >= alpha:
        print "True"
    else:
        print "False"

if __name__ == "__main__":
    main()